internal/server/recon/place_name.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package recon contains code for recon.
+package recon
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+
+	pb "github.com/datacommonsorg/mixer/internal/proto"
+	"github.com/datacommonsorg/mixer/internal/server/v1/propertyvalues"
+	"github.com/datacommonsorg/mixer/internal/store"
+	"github.com/datacommonsorg/mixer/internal/util"
+	"golang.org/x/sync/errgroup"
+	"googlemaps.github.io/maps"
+)
+
+const (
+	maxNumPlacesPerRequest    = 5000
+	maxMapsAPICallsInParallel = 25
+)
+
+// ResolvePlaceNames implements API for Recon.ResolvePlaceNames.
+func ResolvePlaceNames(
+	ctx context.Context,
+	in *pb.ResolvePlaceNamesRequest,
+	store *store.Store,
+	mapsClient *maps.Client,
+) (*pb.ResolvePlaceNamesResponse, error) {
+	if l := len(in.GetPlaces()); l == 0 {
+		return nil, fmt.Errorf("empty input")
+	} else if l > maxNumPlacesPerRequest {
+		return nil, fmt.Errorf(
+			"exceeded max number of places per request (%d): %d",
+			maxNumPlacesPerRequest, l)
+	}
+
+	type placeInfo struct {
+		placeName string
+		placeType string
+	}
+
+	// Load input.
+	nameSet := map[string]struct{}{}
+	placeInfoSet := map[placeInfo]struct{}{}
+	for _, place := range in.GetPlaces() {
+		name := place.GetName()
+		if name == "" {
+			return nil, fmt.Errorf("empty place name")
+		}
+		nameSet[name] = struct{}{}
+		placeInfoSet[placeInfo{name, place.GetType()}] = struct{}{}
+	}
+
+	// Get place IDs.
+	nameToPlaceIDs, placeIDSet, err := resolvePlaceIDs(ctx, mapsClient, nameSet)
+	if err != nil {
+		return nil, err
+	}
+
+	// Resolve place IDs to get DCIDs.
+	placeIDToDCIDs, dcidSet, err := resolveDCIDs(ctx, placeIDSet, store)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get types of the DCIDs.
+	dcidToTypeSet, err := getPlaceTypes(ctx, dcidSet, store)
+	if err != nil {
+		return nil, err
+	}
+
+	// Assemble results.
+	resp := &pb.ResolvePlaceNamesResponse{}
+	for info := range placeInfoSet {
+		place := &pb.ResolvePlaceNamesResponse_Place{
+			Name: info.placeName,
+			Type: info.placeType,
+		}
+
+		if placeIDs, ok := nameToPlaceIDs.Load(info.placeName); ok {
+			allDCIDs := []string{}
+			for _, placeID := range placeIDs.([]string) {
+				if dcids, ok := placeIDToDCIDs[placeID]; ok {
+					allDCIDs = append(allDCIDs, dcids...)
+				}
+			}
+			sort.Strings(allDCIDs)
```

Comment:
Hmm... here it seems this sort might make use lose the order of placeIDs returned from the Maps API?

That might be valuable to keep.  From what I've seen, the first placeId is usually the most relevant one, and subsequent ones may not be right.